From 57892444b383df7c96f93984e265e942d41b0be3 Mon Sep 17 00:00:00 2001
From: Stanley Chu <yschu@nuvoton.com>
Date: Wed, 20 Mar 2024 17:00:53 +0800
Subject: [PATCH 13/16] npcm8xx: Reset I3C controller on probe

I3C controller might not return to idle state after
driver unbind/bind.
Reset controller on driver probe to fix it.

Upstream-Status: Pending [Not submitted to upstream yet]
Signed-off-by: Stanley Chu <yschu@nuvoton.com>
---
 drivers/i3c/master/svc-i3c-master.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/i3c/master/svc-i3c-master.c b/drivers/i3c/master/svc-i3c-master.c
index d897576981ca..71c0873acb33 100644
--- a/drivers/i3c/master/svc-i3c-master.c
+++ b/drivers/i3c/master/svc-i3c-master.c
@@ -17,6 +17,7 @@
 #include <linux/list.h>
 #include <linux/module.h>
 #include <linux/of.h>
+#include <linux/reset.h>
 #include <linux/pinctrl/consumer.h>
 #include <linux/platform_device.h>
 #include <linux/pm_runtime.h>
@@ -1852,6 +1853,7 @@ static int svc_i3c_master_probe(struct platform_device *pdev)
 	struct svc_i3c_master *master;
 	int ret;
 #if defined(CONFIG_ARCH_NPCM)
+	struct reset_control *reset;
 	u32 val;
 #endif
 
@@ -1885,6 +1887,14 @@ static int svc_i3c_master_probe(struct platform_device *pdev)
 	if (ret)
 		return ret;
 
+#if defined(CONFIG_ARCH_NPCM)
+	reset = devm_reset_control_get(&pdev->dev, NULL);
+	if (!IS_ERR(reset)) {
+		reset_control_assert(reset);
+		udelay(5);
+		reset_control_deassert(reset);
+	}
+#endif
 	INIT_WORK(&master->hj_work, svc_i3c_master_hj_work);
 	INIT_WORK(&master->ibi_work, svc_i3c_master_ibi_work);
 	mutex_init(&master->lock);
-- 
2.43.0

